lib/utils/log/file_writer.go

```diff
@@ -0,0 +1,182 @@
+	}
+	fileSharedWriter = &FileSharedWriter{fileFlag: flag, fileMode: mode}
+	fileSharedWriter.logFile.Store(logFile)
+
+	return fileSharedWriter, nil
+}
+
+// CloseFileSharedWriter closes the file shared writer and frees up resources.
+func CloseFileSharedWriter() {
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	if fileSharedWriter != nil {
+		if err := fileSharedWriter.Close(); err != nil {
+			slog.ErrorContext(context.Background(), "Failed to close file shared writer", "error", err)
+		}
+		fileSharedWriter = nil
+	}
+}
+
+// GetFileSharedWriter returns instance of the file shared writer.
+func GetFileSharedWriter() *FileSharedWriter {
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	return fileSharedWriter
+}
+
+// Write writes len(b) bytes from b to the File.
+func (s *FileSharedWriter) Write(b []byte) (int, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	return s.logFile.Load().Write(b)
+}
+
+// Reopen closes the file and opens it again using APPEND mode.
+func (s *FileSharedWriter) Reopen() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	logFile, err := os.OpenFile(s.logFile.Load().Name(), s.fileFlag, s.fileMode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	prevLogFile := s.logFile.Swap(logFile)
+	if err := prevLogFile.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// Close stops watcher loop and frees resources.
+func (s *FileSharedWriter) Close() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if watcher := s.watcher.Swap(nil); watcher != nil {
+		return trace.Wrap(watcher.Close())
+	}
+	s.logFile.Swap(nil)
+
+	return nil
+}
+
+// RunWatcherReopen runs a filesystem watcher for rename/remove events to reopen the log.
+func (s *FileSharedWriter) RunWatcherReopen(ctx context.Context) error {
+	return s.runWatcherFunc(ctx, s.Reopen)
+}
+
+// runWatcherFunc spawns goroutine with the watcher loop to consume events of renaming
+// or removing the log file to trigger the action function when event appeared.
+func (s *FileSharedWriter) runWatcherFunc(ctx context.Context, action func() error) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	context.AfterFunc(ctx, func() {
+		if err := watcher.Close(); err != nil {
+			slog.ErrorContext(context.Background(), "Failed to close file watcher", "error", err)
+		}
+	})
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				if s.logFile.Load().Name() == event.Name && (event.Has(fsnotify.Rename) || event.Has(fsnotify.Remove)) {
+					slog.DebugContext(ctx, "Log file was moved/removed, reopen new one", "file", event.Name)
+					if err := action(); err != nil {
+						slog.ErrorContext(ctx, "Failed to reopen new file", "error", err, "file", event.Name)
+						continue
+					}
+				}
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				slog.ErrorContext(ctx, "Error received on logger watcher", "error", err)
```

Comment:
Do we need to do anything else if we get an error, other than log it?

Reply:
> if we move the log directory somewhere else, we will not able to re-open file since there are no directory, so possible solution it is re-create it

We don't create the directories when we open the log file the first time, right? I'm not sure what the behavior should be in that case, but it seems really awkward to watch for all the directories in the supposed path just to figure out when the file can be reopened, so maybe that's just not our problem and if the directory is moved or deleted we just give up on watching?

Reply:
When we start watching the directory, we actually get event if this specific directory renamed as well, agree with you about creation since we don't know parameter to create it, just was suggestion. If directory is moved `fsnotify` removes this path from watch list

Reply:
Can we pick the initial fsmode of the parent directory and create it using the same value?

Reply:
@tigrato it might unpredictable behavior, since we keep writing in the moved file with directory (reopen fails) we should be good just to raise an error